wvd-templates/DSC/Functions.ps1

```diff
@@ -222,13 +222,159 @@ function ExtractDeploymentAgentZipFile {
+        [AllowEmptyString()]
+        [string]$TenantId = ""
+    )
+    
+    if ($ServicePrincipal) {
+        Write-Log -Message "Authenticating using service principal $($Credential.username) and Tenant id: $TenantId "
+    }
+    else {
+        $PSBoundParameters.Remove('ServicePrincipal')
+        $PSBoundParameters.Remove('TenantId')
+        Write-Log -Message "Authenticating using user $($Credential.username)"
+    }
+
+    $authentication = $null
+    try {
+        $authentication = Add-RdsAccount @PSBoundParameters
+        if (!$authentication) {
+            throw $authentication
+        }
+    }
+    catch {
+        $errMsg = "Windows Virtual Desktop Authentication Failed, Error:`n$($_ | Out-String)"
+        Write-Log -Error "$errMsg"
+        throw "$errMsg"
+    }
+    Write-Log -Message "Windows Virtual Desktop Authentication successfully Done. Result:`n$($authentication | Out-String)"
+}
+
+function SetTenantContextAndValidate {
+    param(
+        [Parameter(mandatory = $true)]
+        [string]$definedTenantGroupName,
+
+        [Parameter(mandatory = $true)]
+        [string]$TenantName
+    )
+    #//todo refactor
+    #//todo try catch ?
+    # Set context to the appropriate tenant group
+    $currentTenantGroupName = (Get-RdsContext).TenantGroupName
+    if ($definedTenantGroupName -ne $currentTenantGroupName) {
+        Write-Log -Message "Running switching to the $definedTenantGroupName context"
+        Set-RdsContext -TenantGroupName $definedTenantGroupName
+    }
+    try {
+        $tenants = Get-RdsTenant -Name "$TenantName"
+        if (!$tenants) {
+            Write-Log "No tenants exist or you do not have proper access."
+            #//todo throw ?
+        }
+    }
+    catch {
+        #//todo refactor msg ?
+        Write-Log -Message $_
+        throw $_
+    }
+}
+
+function ImportRDPSMod {
+    param(
+        [string]$Source = 'attached',
+        [string]$ArtifactsPath
+
+        # [Parameter(mandatory = $false)]
+        # [string]$DeployAgentLocation = 'C:\DeployAgent'
+    )
+
+    #//todo move outside
+    # Write-Log -Message "Creating a folder inside rdsh vm for extracting deployagent zip file"
+    # ExtractDeploymentAgentZipFile -ScriptPath $ScriptPath -DeployAgentLocation $DeployAgentLocation
+    
+    # Write-Log -Message "Changing current folder to Deployagent folder: $DeployAgentLocation"
+    # Set-Location "$DeployAgentLocation"
+    
+    if ($Source -eq 'attached') {
+        if ((-not $ArtifactsPath) -or (-not (test-path $ArtifactsPath))) {
+            throw "invalid param: ArtifactsPath = '$ArtifactsPath'"
+        }
+
+        # Locating and extracting PowerShellModules.zip
+        $ZipName = 'PowerShellModules.zip'
+        Write-Log -Message "Locating '$ZipName' within Custom Script Extension folder structure: '$ArtifactsPath'"
+        $ZipPath = (Get-ChildItem "$ArtifactsPath\" -Filter $ZipName -Recurse | Select-Object -First 1).FullName

```

Comment:
Instead of only selecting the first item it seems like this should throw an exception if more than one "PowerShellModules.zip" is found since that is not expected.